File: src/main/java/com/openlattice/chronicle/controllers/legacy/ChronicleStudyController.java
/*
 * Copyright (C) 2018. OpenLattice, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 */

package com.openlattice.chronicle.controllers.legacy;

import com.codahale.metrics.annotation.Timed;
import com.google.common.base.Optional;
import com.openlattice.chronicle.ChronicleStudyApi;
import com.openlattice.chronicle.data.ChronicleAppsUsageDetails;
import com.openlattice.chronicle.data.ChronicleQuestionnaire;
import com.openlattice.chronicle.data.ParticipationStatus;
import com.openlattice.chronicle.services.enrollment.EnrollmentManager;
import com.openlattice.chronicle.services.surveys.SurveysManager;
import com.openlattice.chronicle.sources.Datasource;
import org.apache.olingo.commons.api.edm.FullQualifiedName;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import javax.inject.Inject;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

/**
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
@RestController
@RequestMapping( ChronicleStudyApi.CONTROLLER )
public class ChronicleStudyController implements ChronicleStudyApi {

    @Inject
    private SurveysManager surveysManager;

    @Inject
    private EnrollmentManager enrollmentManager;

    @Override
    @Timed
    @RequestMapping(
            path = STUDY_ID_PATH + PARTICIPANT_ID_PATH + DATASOURCE_ID_PATH,
            method = RequestMethod.POST,
            consumes = MediaType.APPLICATION_JSON_VALUE
    )
    public UUID enrollSource(
            @PathVariable( STUDY_ID ) UUID studyId,
            @PathVariable( PARTICIPANT_ID ) String participantId,
            @PathVariable( DATASOURCE_ID ) String datasourceId,
            @RequestBody Optional<Datasource> datasource ) {

        return enrollmentManager.registerDataSource( null, studyId, participantId, datasourceId, datasource );
    }

    @Override
    @Timed
    @RequestMapping(
            path = STUDY_ID_PATH + PARTICIPANT_ID_PATH + DATASOURCE_ID_PATH,
            method = RequestMethod.GET,
            produces = MediaType.APPLICATION_JSON_VALUE
    )
    public Boolean isKnownDatasource(
            @PathVariable( STUDY_ID ) UUID studyId,
            @PathVariable( PARTICIPANT_ID ) String participantId,
            @PathVariable( DATASOURCE_ID ) String datasourceId ) {
        //  validate that this device belongs to this participant in this study
        //  look up in association entitySet between device and participant, and device and study to see if it exists?
        //  DataApi.getEntity(entitySetId :UUID, entityKeyId :UUID)
        return enrollmentManager.isKnownDatasource( null, studyId, participantId, datasourceId );
    }

    @Timed
    @Override
    @RequestMapping(
            path = PARTICIPANT_PATH + DATA_PATH + STUDY_ID_PATH + PARTICIPANT_ID_PATH + APPS,
            method = RequestMethod.GET,
            produces = MediaType.APPLICATION_JSON_VALUE
    )
    public List<ChronicleAppsUsageDetails> getParticipantAppsUsageData(
            @PathVariable( STUDY_ID ) UUID studyId,
            @PathVariable( PARTICIPANT_ID ) String participantId,
            @RequestParam( value = DATE ) String date ) {
        return surveysManager.getParticipantAppsUsageData( null, studyId, participantId, date );
    }

    @Override
    @Timed
    @RequestMapping(
            path = STUDY_ID_PATH + NOTIFICATIONS,
            method = RequestMethod.GET
    )
    public Boolean isNotificationsEnabled(
            @PathVariable( STUDY_ID ) UUID studyId ) {
        return enrollmentManager.isNotificationsEnabled( null, studyId );
    }

    @Override
    @Timed
    @RequestMapping(
            path = STUDY_ID_PATH + PARTICIPANT_ID_PATH + ENROLLMENT_STATUS,
            method = RequestMethod.GET
    )
    public ParticipationStatus getParticipationStatus(
            @PathVariable( STUDY_ID ) UUID studyId,
            @PathVariable( PARTICIPANT_ID ) String participantId ) {
        return enrollmentManager.getParticipationStatus( null, studyId, participantId );
    }

    @Override
    @RequestMapping(
            path = PARTICIPANT_PATH + DATA_PATH + STUDY_ID_PATH + PARTICIPANT_ID_PATH + APPS,
            method = RequestMethod.POST,
            consumes = MediaType.APPLICATION_JSON_VALUE
    )
    public void submitAppUsageSurvey(
            @PathVariable( STUDY_ID ) UUID studyId,
            @PathVariable( PARTICIPANT_ID ) String participantId,
            @RequestBody Map<UUID, Map<FullQualifiedName, Set<Object>>> associationDetails ) {

        surveysManager.submitAppUsageSurvey( null, studyId, participantId, associationDetails );
    }

    @Override
    @Timed
    @RequestMapping(
            path = STUDY_ID_PATH + QUESTIONNAIRE + ENTITY_KEY_ID_PATH,
            method = RequestMethod.GET,
            produces = MediaType.APPLICATION_JSON_VALUE
    )
    public ChronicleQuestionnaire getChronicleQuestionnaire(
            @PathVariable( STUDY_ID ) UUID studyId,
            @PathVariable( ENTITY_KEY_ID ) UUID questionnaireEKID
    ) {
        return surveysManager.getQuestionnaire( null, studyId, questionnaireEKID );
    }

    @Override
    @Timed
    @RequestMapping(
            path = STUDY_ID_PATH + PARTICIPANT_ID_PATH + QUESTIONNAIRE,
            method = RequestMethod.POST,
            consumes = MediaType.APPLICATION_JSON_VALUE
    )
    public void submitQuestionnaire(
            @PathVariable( STUDY_ID ) UUID studyId,
            @PathVariable( PARTICIPANT_ID ) String participantId,
            @RequestBody Map<UUID, Map<FullQualifiedName, Set<Object>>> questionnaireResponses ) {

        surveysManager.submitQuestionnaire( null, studyId, participantId, questionnaireResponses );
    }

    @Override
    @Timed
    @RequestMapping(
            path = STUDY_ID_PATH + QUESTIONNAIRES,
            method = RequestMethod.GET,
            produces = MediaType.APPLICATION_JSON_VALUE
    )
    public Map<UUID, Map<FullQualifiedName, Set<Object>>> getStudyQuestionnaires(
            @PathVariable( STUDY_ID ) UUID studyId ) {

        return surveysManager.getStudyQuestionnaires( null, studyId );
    }

    @RequestMapping(
            path = STUDY_ID_PATH + PARTICIPANT_ID_PATH + TIME_USE_DIARY,
            method = RequestMethod.POST,
            consumes = MediaType.APPLICATION_JSON_VALUE
    )
    @Override
    public void submitTimeUseDiarySurvey(
            @PathVariable ( STUDY_ID ) UUID studyId,
            @PathVariable ( PARTICIPANT_ID ) String participantId,
            @RequestBody List<Map<FullQualifiedName, Set<Object>>> surveyData
    ) {
        surveysManager.submitTimeUseDiarySurvey( null, studyId, participantId, surveyData );
    }
}

File: src/main/kotlin/com/openlattice/chronicle/controllers/legacy/ChronicleStudyController.kt
package com.openlattice.chronicle.controllers.legacy

import com.codahale.metrics.annotation.Timed
import com.google.common.base.Optional
import com.openlattice.chronicle.ChronicleStudyApi
import com.openlattice.chronicle.data.ParticipationStatus
import com.openlattice.chronicle.participants.ParticipantStats
import com.openlattice.chronicle.services.enrollment.EnrollmentManager
import com.openlattice.chronicle.services.studies.StudyService
import com.openlattice.chronicle.services.surveys.SurveysManager
import com.openlattice.chronicle.sources.AndroidDevice
import com.openlattice.chronicle.sources.IOSDevice
import com.openlattice.chronicle.sources.SourceDevice
import org.apache.olingo.commons.api.edm.FullQualifiedName
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RestController
import java.time.OffsetDateTime
import java.util.UUID
import javax.inject.Inject

/**
 *
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
@RestController
@RequestMapping(ChronicleStudyApi.CONTROLLER)
class ChronicleStudyController : ChronicleStudyApi {
    @Inject
    private lateinit var surveysManager: SurveysManager

    @Inject
    private lateinit var enrollmentManager: EnrollmentManager

    @Inject
    private lateinit var studyService: StudyService

    @Timed
    @RequestMapping(
        path = [ChronicleStudyApi.STUDY_ID_PATH + ChronicleStudyApi.PARTICIPANT_ID_PATH + ChronicleStudyApi.DATASOURCE_ID_PATH],
        method = [RequestMethod.POST], consumes = [MediaType.APPLICATION_JSON_VALUE]
    )
    override fun enrollSource(
        @PathVariable(ChronicleStudyApi.STUDY_ID) studyId: UUID,
        @PathVariable(ChronicleStudyApi.PARTICIPANT_ID) participantId: String,
        @PathVariable(ChronicleStudyApi.DATASOURCE_ID) datasourceId: String,
        @RequestBody sourceDevice: Optional<SourceDevice>
    ): UUID {
        val realStudyId = studyService.getStudyId(studyId)
        checkNotNull(realStudyId) { "invalid study id" }
        val id = enrollmentManager.registerDatasource(realStudyId, participantId, datasourceId, sourceDevice.get())
        studyService.updateLastDevicePing(realStudyId, participantId, sourceDevice.get())
        return id
    }

    @Timed
    @RequestMapping(
        path = [ChronicleStudyApi.STUDY_ID_PATH + ChronicleStudyApi.PARTICIPANT_ID_PATH + ChronicleStudyApi.DATASOURCE_ID_PATH],
        method = [RequestMethod.GET], produces = [MediaType.APPLICATION_JSON_VALUE]
    )
    override fun isKnownDatasource(
        @PathVariable(ChronicleStudyApi.STUDY_ID) studyId: UUID,
        @PathVariable(ChronicleStudyApi.PARTICIPANT_ID) participantId: String,
        @PathVariable(ChronicleStudyApi.DATASOURCE_ID) datasourceId: String
    ): Boolean {
        val realStudyId = studyService.getStudyId(studyId)
        checkNotNull(realStudyId) { "invalid study id" }
        return enrollmentManager.isKnownDatasource(realStudyId, participantId, datasourceId)
    }

    @Timed
    @RequestMapping(
        path = [ChronicleStudyApi.STUDY_ID_PATH + ChronicleStudyApi.NOTIFICATIONS], method = [RequestMethod.GET]
    )
    override fun isNotificationsEnabled(
        @PathVariable(ChronicleStudyApi.STUDY_ID) studyId: UUID
    ): Boolean {
        val realStudyId = studyService.getStudyId(studyId)
        checkNotNull(realStudyId) { "invalid study id" }
        return studyService.isNotificationsEnabled(realStudyId)
    }

    @Timed
    @RequestMapping(
        path = [ChronicleStudyApi.STUDY_ID_PATH + ChronicleStudyApi.PARTICIPANT_ID_PATH + ChronicleStudyApi.ENROLLMENT_STATUS],
        method = [RequestMethod.GET]
    )
    override fun getParticipationStatus(
        @PathVariable(ChronicleStudyApi.STUDY_ID) studyId: UUID,
        @PathVariable(ChronicleStudyApi.PARTICIPANT_ID) participantId: String
    ): ParticipationStatus {
        val realStudyId = studyService.getStudyId(studyId)
        checkNotNull(realStudyId) { "invalid study id" }
        return enrollmentManager.getParticipationStatus(realStudyId, participantId)
    }

    @Timed
    @RequestMapping(
        path = [ChronicleStudyApi.STUDY_ID_PATH + ChronicleStudyApi.QUESTIONNAIRES], method = [RequestMethod.GET],
        produces = [MediaType.APPLICATION_JSON_VALUE]
    )
    override fun getStudyQuestionnaires(
        @PathVariable(ChronicleStudyApi.STUDY_ID) studyId: UUID
    ): Map<UUID, Map<FullQualifiedName, Set<Any>>> {
        val organizationId = studyService.getOrganizationIdForLegacyStudy(studyId)
        val realStudyId = studyService.getStudyId(studyId)
        checkNotNull(realStudyId) { "invalid study id" }
        return surveysManager.getLegacyStudyQuestionnaires(organizationId, realStudyId)
    }
}
